File: 704.binary-search.py
#
# @lc app=leetcode id=704 lang=python3
#
# [704] Binary Search
#

# @lc code=start
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        start = 0
        end = len(nums) - 1

        while start <= end:
            mid = (start + end)//2
            if target < nums[mid]:
                end = mid - 1
            elif target > nums[mid]:
                start = mid + 1
            else:
                return mid

        return -1    
# @lc code=end

File: 704.binary-search.rs
/*
 * @lc app=leetcode id=704 lang=rust
 *
 * [704] Binary Search
 */

// @lc code=start
impl Solution {
    pub fn search(nums: Vec<i32>, target: i32) -> i32 {
        let mut begin = 0;
        let mut end = nums.len() - 1;

        while begin < end {
            let mid = begin + (end - begin) / 2;
            if target < nums[mid] {
                end = mid - 1;
            } else if target > nums[mid] {
                begin = mid + 1;
            } else {
                return mid as i32;
            }
        }

        return -1;
    }
}

// @lc code=end
